awpy/analytics/map_control.py

```diff
@@ -0,0 +1,413 @@
+"""Functions for calculating map control values and metrics.
+
+    A team's map control can be thought of as the sum of it's
+    individual player's control.
+
+    Example notebook:
+
+    # pylint: disable=line-too-long
+    github.com/pnxenopoulos/awpy/blob/main/examples/05_Map_Control_Calculations_And_Visualizations.ipynb
+"""
+
+from collections import defaultdict, deque
+
+import numpy as np
+
+from awpy.analytics.nav import area_distance, calculate_tile_area, find_closest_area
+from awpy.data import NAV, NAV_GRAPHS
+from awpy.types import (
+    BFSTileData,
+    FrameMapControlValues,
+    FrameTeamMetadata,
+    GameFrame,
+    GameRound,
+    PlayerPosition,
+    TeamFrameInfo,
+    TeamMapControlValues,
+    TeamMetadata,
+    TileDistanceObject,
+    TileId,
+    TileNeighbors,
+)
+
+
+def _approximate_neighbors(
+    map_name: str,
+    source_tile_id: TileId,
+    n_neighbors: int = 5,
+) -> list[TileDistanceObject]:
+    """Approximates neighbors for isolated tiles by finding n closest tiles.
+
+    Args:
+        map_name (str): Map for source_tile_id
+        source_tile_id (TileId): TileId for source tile
+        n_neighbors (int): Number of closest tiles/approximated neighbors wanted
+
+    Returns:
+        List of TileDistanceObjects for n closest tiles
+
+    Raises:
+        ValueError: If source_tile_id is not in awpy.data.NAV[map_name]
+                    If n_neighbors <= 0
+    """
+    if source_tile_id not in NAV[map_name]:
+        msg = "Tile ID not found."
+        raise ValueError(msg)
+    if n_neighbors <= 0:
+        msg = "Invalid n_neighbors value. Must be > 0."
+        raise ValueError(msg)
+
+    current_map_info = NAV[map_name]
+    possible_neighbors_arr: list[TileDistanceObject] = []
+
+    for tile in current_map_info:
+        if tile != source_tile_id:
+            current_tile_distance_obj = TileDistanceObject(
+                tile_id=tile,
+                distance=area_distance(
+                    map_name, tile, source_tile_id, dist_type="euclidean"
+                )["distance"],
+            )
+
+            possible_neighbors_arr.append(current_tile_distance_obj)
+
+    return sorted(possible_neighbors_arr, key=lambda d: d.distance)[:n_neighbors]
+
+
+def _bfs(
+    map_name: str,
+    current_tiles: list[TileId],
+    neighbor_info: TileNeighbors,
+    max_depth: int = 10,
```

Comment:
similar as above -- do you think this default is globally ok? are there some maps where it might not work well?

Reply:
I think 10 is okay as a default at least for now. One reason I kept it at 10 was that there wasn't too many tiles to search for the bfs algorithm. And visually the spread of 10 looked to take up a reasonable amount of space for each player on the maps I took a look at (mirage and inferno mainly I believe).

Here are a few ideas I came up with so we could try to address this in the future, let me know what you think:
1. Change the max_depth based on the size of the map requested
2. Use max_area instead of a max_depth (so keep a running count of the area of the tiles processed already and once it reaches a threshold we stop the algorithm)
3. Don't have a default value and have the user pass it in

Reply:
Like above, do some sensitivity testing just to see. It will be useful for your own work with map control, too

Reply:
I actually really like the default area idea. Because especially if you take a look at anubis or overpass stuff could get weird in certain places.
![tiles_de_anubis](https://github.com/pnxenopoulos/awpy/assets/47750513/adf23931-abc0-420b-a80d-35bad261877a)

Reply:
Yeah I think going to area would be better especailly after seeing ancient. Any idea what would be a good way to decide a default max_area threshold? I was thinking of experimenting with 1/10 of 1/20 of the average area across all maps but there is likely a better way to find a good default.